### ISSUE TITLE
ImportError: Cannot load backend 'TkAgg' which requires the 'tk' interactive framework, as 'headless' is currently running

### ISSUE THREAD

**Issue by kk6398** [muskie82/MonoGS#38]
A error occurs when  “ python slam.py --config configs/mono/tum/fr3_office.yaml” 

![1711281066057](https://github.com/muskie82/MonoGS/assets/70420660/3f5dec5f-44bc-491d-ae24-54c6b3dbd35a)

**Comment by kk6398** [comment]
I use SSH to remotely connect to the Linux terminal. The vscode shell I use does not have a graphical interface. Is this the reason?

**Comment by kk6398** [comment]
Yep, I try run "python slam.py --config configs/mono/tum/fr3_office.yaml" on the unbuntu graphical interface, and the error "Cannot load backend 'TkAgg' which requires the 'tk' interactive framework, as 'headless' is currently running" doesnot appear.  

However, anthor question occurs: ①there is no gui window pop out, and ②"cuda out of memory" when processing"Evaluating ATE at frame:  2230".     (I run the code on 4090 24G)    

What is the reason and how I deal with it?

**Comment by muskie82** [comment]
Hi,

Never seen this before, but I guess the GUI is dead for some reason and the q_main2vis is not cleaned up, which leads to out of memory.
Do you see any suspicious error message related to GUI?

**Comment by kk6398** [comment]
okay, I display the code log as follows:
`python slam.py --config configs/mono/tum/fr1_desk.yaml 
MonoGS: saving results in 
results/tum_rgbd_dataset_freiburg1_desk/2024-03-24-21-07-31
FEngine (64 bits) created at 0x53785400 (threading is enabled)
FEngine resolved backend: OpenGL
X Error of failed request:  GLXBadFBConfig
  Major opcode of failed request:  152 (GLX)
  Minor opcode of failed request:  34 ()
  Serial number of failed request:  39
  Current serial number in output stream:  37
MonoGS: Resetting the system
MonoGS: Initialized map
MonoGS: Resetting the opacity of non-visible Gaussians
MonoGS: Keyframes lacks sufficient overlap to initialize the map, resetting.
MonoGS: Resetting the system
MonoGS: Initialized map
MonoGS: Resetting the opacity of non-visible Gaussians
MonoGS: Performing initial BA for initialization
MonoGS: Initialized SLAM
MonoGS: Evaluating ATE at frame:  74
Eval: RMSE ATE [m] 0.008595195845623836
MonoGS: Evaluating ATE at frame:  123
Eval: RMSE ATE [m] 0.01270222476891415
MonoGS: Evaluating ATE at frame:  164
Eval: RMSE ATE [m] 0.01736790069366283
MonoGS: Resetting the opacity of non-visible Gaussians
MonoGS: Evaluating ATE at frame:  203
Eval: RMSE ATE [m] 0.022410050076178276
MonoGS: Evaluating ATE at frame:  257
Eval: RMSE ATE [m] 0.029316772769155855
MonoGS: Evaluating ATE at frame:  294
Eval: RMSE ATE [m] 0.033102492927375486
MonoGS: Evaluating ATE at frame:  323
Eval: RMSE ATE [m] 0.03582281334425646
MonoGS: Resetting the opacity of non-visible Gaussians
MonoGS: Evaluating ATE at frame:  367
Eval: RMSE ATE [m] 0.03840116612275182
MonoGS: Evaluating ATE at frame:  399
Eval: RMSE ATE [m] 0.0379590706327925
MonoGS: Evaluating ATE at frame:  451
Eval: RMSE ATE [m] 0.03892031242012154
MonoGS: Evaluating ATE at frame:  494
Eval: RMSE ATE [m] 0.03823152505622439
MonoGS: Resetting the opacity of non-visible Gaussians
MonoGS: Evaluating ATE at frame:  525
Eval: RMSE ATE [m] 0.038492287772041056
Eval: RMSE ATE [m] 0.03811678444826245
Eval: Total time 462.99578125
Eval: Total FPS 1.2786293611611608
MonoGS: Backend stopped and joined the main thread
MonoGS: GUI Stopped and joined the main thread
MonoGS: Done.
`

**Comment by kk6398** [comment]
okay, I display the code log as follows:
python slam.py --config configs/mono/tum/fr1_desk.yaml
![image](https://github.com/muskie82/MonoGS/assets/70420660/59962658-b26e-47d6-a431-d358fe32dca6)
![image](https://github.com/muskie82/MonoGS/assets/70420660/7a537b54-1d99-4cb4-8efb-44fc90d9bd8e)

**Comment by muskie82** [comment]
I think `x error of failed request` is the one related.
Your environment seeems to have some issue with OpenGL, which prevents GUI launch.

Can you try to fix the error message?
As a quick search, this might be relevant.
https://askubuntu.com/questions/1369900/x-error-of-failed-request-glxbadfbconfig

**Comment by kk6398** [comment]
Thank for your reply. I tried entering "export MESA_GL_VERSION_OVERRIDE=4.3" or export "MESA_GL_VERSION_OVERRIDE=4.5" as the ISSUE#5, but the window MonoGS only flashed for one second and suddenly exiting, then reporting an error as follow. I hope you can answer.   The version of pyopengl is 3.1.7   cry~ ~    
![image](https://github.com/muskie82/MonoGS/assets/70420660/721ecc92-db62-43e5-b97f-819bf741218d)

**Comment by kk6398** [comment]
> Thank for your reply. I tried entering "export MESA_GL_VERSION_OVERRIDE=4.3" or export "MESA_GL_VERSION_OVERRIDE=4.5" as the ISSUE#5, but the window MonoGS only flashed for one second and suddenly exiting, then reporting an error as follow. I hope you can answer. The version of pyopengl is 3.1.7 cry~ ~ ![image](https://private-user-images.githubusercontent.com/70420660/316535671-721ecc92-db62-43e5-b97f-819bf741218d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzNzc4MTksIm5iZiI6MTcxMTM3NzUxOSwicGF0aCI6Ii83MDQyMDY2MC8zMTY1MzU2NzEtNzIxZWNjOTItZGI2Mi00M2U1LWI5N2YtODE5YmY3NDEyMThkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI1VDE0MzgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzYTNlNDRmNDAwNjE3ZmIxZWQ4NTQwYzRhYWNlMzRkODM4OThlYjkzMzRhOTc4YWM4YWNjYWJhYmJjYWFlYTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1OfVTyYjxXiJ-8EEQlvuoC-InGwg-u0J7LtSM6ASk-Q)

I may not have fully stated the environment:   ubuntu 18.04   GPU: 4090   cuda11.6  pytorch1.12.1   pyopengl 3.1.7

**Comment by muskie82** [comment]
I don't have an immediate answer unfortunately. The error message `x error of failed request` has many reports online, so I recommend searching the web for similar cases and trying out the suggested solutions.

For a temporary solution, you could comment out all the code related to `gl_render` in `slam_gui.py`. `gl_render` is used just for the Ellipsoid shader, so the rest of the GUI should function without it.

**Comment by kk6398** [comment]
Hi, I try lots of methods, and successfully run the MonoGS GUI window through [https://github.com/isl-org/Open3D/issues/4429], but some errors occur: there is no picture appears in the left frame. Moreover, there is no response when l click any button on the right.
![image](https://github.com/muskie82/MonoGS/assets/70420660/25577e44-0e75-4ef8-af44-4d3fb0863a0c)

**Comment by muskie82** [comment]
Any error message?

Also, I recommend you to try some of the Open3D GUI examples, for example [this](https://github.com/isl-org/Open3D/blob/main/examples/python/visualization/demo_scene.py). If the same problem happens, then probably some similar issues/solutions are reported on Open3D repo.

**Comment by yaochengrong** [comment]
same issue here

**Comment by kk6398** [comment]
> same issue here

I have not solved it yet. Do you have any good methods?

**Comment by peterWon** [comment]
@kk6398 I met this error. I solved this problem by 'evo_config set plot_backend agg'

**Comment by kk6398** [comment]
> @kk6398 I met this error. I solved this problem by 'evo_config set plot_backend agg'

Can you explain the operation steps in detail please?

**Comment by whwh747** [comment]
i meet the same error in vscode ， and the i run this code in mobaxterm ， it is normal @kk6398

**Comment by kk6398** [comment]
> i meet the same error in vscode ， and the i run this code in mobaxterm ， it is normal @kk6398

Thank for your reply. I have run this code in mobaxterm.  However, the real-time performance of SLAM displayed in the pop-up window is poor, and it does not perform tracking and mapping as quickly as the video provided by the author.

**Comment by whwh747** [comment]
i agree , but i did not find the reason, and my code seems to get stuck after running
![00ab49f29cbcf9956de56fd1ff3bda3](https://github.com/muskie82/MonoGS/assets/63688127/f853f2e9-1d2f-457f-a8e4-1f636a5691d5)
@kk6398

**Comment by rmurai0610** [comment]
To solve 
> Cannot load backend 'TkAgg' which requires the 'tk' interactive framework, as 'headless' is currently running

As @peterWon mentions, if you run the command in the terminal,
`evo_config set plot_backend agg`
agg will be set for the backend for the evo. (evo has a global config file, and it's probably set to TkAgg for some reason)


@kk6398 @whwh747, which sequences are you running? 
@whwh747 I cannot replicate the 100% CPU utilisation. What CPU are you using?

**Comment by whwh747** [comment]
rgbd model fr3 long office ， and you can clearly see the CPU usage and which processes are using the CPU in nvitop. @rmurai0610 
<img width="319" alt="885d49fdb219aa01f47457390b924ee" src="https://github.com/muskie82/MonoGS/assets/63688127/b05ea52e-3802-4ef1-9a72-cff583622da0">
